osu.Game/Graphics/Containers/BeatSyncedContainer.cs

```diff
@@ -91,7 +91,17 @@ protected override void Update()
 
             if (IsBeatSyncedWithTrack)
             {
-                currentTrackTime = BeatSyncSource.Clock.CurrentTime + EarlyActivationMilliseconds;
+                double early = EarlyActivationMilliseconds;
+
+                // In the case of gameplay, we are usually within a hierarchy with the correct rate applied to our `Drawable.Clock`.
+                // This means that the amount of early adjustment is adjusted in line with audio track rate changes.
+                // But other cases like the osu! logo at the main menu won't correctly have this rate information.
+                //
+                // So for cases where the rate of the source isn't in sync with our hierarchy, let's assume we need to account for it locally.
+                if (Clock.Rate == 1 && BeatSyncSource.Clock.Rate != Clock.Rate)
```

Comment:
The general idea here appears to check out but I am left wondering about the `Clock.Rate == 1` check here. Is this not an issue for any clock rate mismatch? And as such the early activation period should be adjusted by a ratio of `BeatSyncSource.Clock.Rate / Clock.Rate`?

Reply:
I thought I replied to this way back but it must have been in my dreams, or lost as a draft..

Re-reading the comment above, I think what I have makes sense. It's only aiming to apply a correction when it's very clear that we haven't already taken on a hierarchical rate adjustment (because if we have, it's assumed that things are already in a good state).

Your proposal may work though, since in the case we *have* got a hierarchical rate, it would be `early *= 1`.

I think your proposal is to change this to:

```diff
diff --git a/osu.Game/Graphics/Containers/BeatSyncedContainer.cs b/osu.Game/Graphics/Containers/BeatSyncedContainer.cs
index a14dfd4d64..de9a5aff3a 100644
--- a/osu.Game/Graphics/Containers/BeatSyncedContainer.cs
+++ b/osu.Game/Graphics/Containers/BeatSyncedContainer.cs
@@ -96,10 +96,8 @@ protected override void Update()
                 // In the case of gameplay, we are usually within a hierarchy with the correct rate applied to our `Drawable.Clock`.
                 // This means that the amount of early adjustment is adjusted in line with audio track rate changes.
                 // But other cases like the osu! logo at the main menu won't correctly have this rate information.
-                //
-                // So for cases where the rate of the source isn't in sync with our hierarchy, let's assume we need to account for it locally.
-                if (Clock.Rate == 1 && BeatSyncSource.Clock.Rate != Clock.Rate)
-                    early *= BeatSyncSource.Clock.Rate;
+                // We can adjust here to ensure the applied early activation always matches expectations.
+                early *= BeatSyncSource.Clock.Rate / Clock.Rate;
 
                 currentTrackTime = BeatSyncSource.Clock.CurrentTime + early;
 

```

?

Reply:
That would be the proposal I think yeah.

Reply:
Seems to work well enough, have applied.

Reply:
https://github.com/ppy/osu/pull/26707/commits/e02ad6cf94d3a9bba0cbcac6ecee5a0125f9da89 has me thinking that maybe a guard against zero rate is warranted somewhere closer to this change rather than doing test changes...

Reply:
dunno. i don't think we should ever be hitting zero rate anymore. your call though.

Reply:
I've pushed a guard (https://github.com/ppy/osu/pull/26707/commits/0b29a762b8aa712d1a0b697f367ae809a0896f15) because I have the unshakable feeling that if we don't now we're gonna have to hotfix this later.